### PR TITLE
Test, implement & integrate: LRU strategy for test runs in searchcache

### PR DESCRIPTION
**Pull request by mdittmer**
Currently `searchcache` evicts runs according to their "time started". This change usees an LRU strategy for evicting test runs instead.

### CODE REVIEW

File: api/query/cache/lru/lru.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+func (l *lru) syncEvictLRU() int64 {
+	l.m.Lock()
+	defer l.m.Unlock()
+
+	rs := make(byTime, 0, len(l.byRunID))
+	for r, t := range l.byRunID {
+		rs = append(rs, lruEntry{r, t})
+	}
+	sort.Sort(rs)
```
[the remark this answers is not in the document]
Comment (mdittmer) [reply api/query/cache/lru/lru.go]: In principle, the old and the new methods could be implemented `O(n)`. In practice, both used sort-and-take-first. I can change this to be `O(n)`, though; that's better :smile:

**Comment by mdittmer** [comment]
@Hexcles pointed out something important: With the `EvictLRU()` => `EvictAnyRun()` flow, we pay the cost of identifying and evicting one LRU run frequently. (Presumably, the next time a run is added to the cache it will hit the soft memory limit again.)

This change now reflects the first step in moving to a `Monitor` that is bound to a percentage of runs to evict when the soft memory limit is hit. The percentage will be passed along to `EvictLRU(percent)`, implemented here. In this scheme, a single critical section pays the cost of sorting runs and then evicting a batch of them.

**Comment by mdittmer** [comment]
_ping_ @lukebjerring

File: api/query/cache/lru/lru.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+type lruEntry struct {
+	int64
+	time.Time
+}
+
+type byTime []lruEntry
+
+func (s byTime) Len() int {
```
Comment (lukebjerring) [inline api/query/cache/lru/lru.go]: nit: I think `gofmt` lets these 3 `sort` methods be one-liners

File: api/query/cache/lru/lru.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukebjerring) [reply api/query/cache/lru/lru.go]: For a fraction eviction this is no longer an issue.

**Comment by mdittmer** [comment]
Currently `searchcache` evicts runs according to their "time started". This change usees an LRU strategy for evicting test runs instead.

**Pull request merged**